Libraries/CLI/src/cli.h

```diff
@@ -0,0 +1,155 @@
+ * trademarks, maskwork rights, or any other form of intellectual
+ * property whatsoever. Maxim Integrated Products, Inc. retains all
+ * ownership rights.
+ *
+ ******************************************************************************/
+
+/*! \file cli.h
+    \brief A CLI Implementation using a command table to store command string, function pointer and a help string is dispatched.
+
+    Details. 
+
+    1.  Line Accumlator 
+
+    Reads the incoming bytes
+    Accumulates into a line buffer
+    Echos char back to the emulator
+    Handles backspace
+
+    2.  Prcoess Command
+
+    Processes input into a series of tokens
+    All tokes are seperated by whitespace characters 
+    Lookup first token in a table of functions
+    Dispatch to handler functions 
+    
+*/
+
+/* -------------------------------------------------- */
+//                 INCLUDE GUARD
+/* -------------------------------------------------- */
+#ifndef EXAMPLES_MAX78000_SDHC_FTHR_INCLUDE_CLI_H_
+#define EXAMPLES_MAX78000_SDHC_FTHR_INCLUDE_CLI_H_
+
+/* -------------------------------------------------- */
+//                      INCLUDES
+/* -------------------------------------------------- */
+#include <stdio.h>
+#include <stdlib.h>
+#include <ctype.h>
+#include <stdbool.h>
+#include <stddef.h>
+#include <stdint.h>
+#include <string.h>
+#include "board.h"
+#include "led.h"
+#include "mxc_device.h"
+#include "mxc_errors.h"
+#include "uart.h"
+#include "user-cli.h"
+
+/* -------------------------------------------------- */
+//                      MACROS
+/* -------------------------------------------------- */
+#define ENTER                   0X0D
+#define NEW_LINE                0x0A
+#define SPACE                   0x20
+#define TAB                     0x09
+#define BACKSPACE               0X08
+#define MAXBUFF                 2000
+#define DELETE                  0x7F
+#define DOLLAR                  0x24
+
+
+#define ARROW_KEY_CODE_1        0x1B
+#define ARROW_KEY_CODE_2        0x5B
+#define ARROW_KEY_CODE_LEFT     0x44
+#define ARROW_KEY_CODE_RIGHT    0x43
+#define ARROW_KEY_CODE_UP       0x41
+#define ARROW_KEY_CODE_DOWN     0x42
+
+
+/* -------------------------------------------------- */
+//                 FUNCTION PROTOTYPES
+/* -------------------------------------------------- */
+
+/**
+ * @brief Reads incoming bytes, Accumulate into line buffer, Echo's chars back to the other side and handles backspace.
+ * 		  It calls the process_command function upon pressing the ENTER key
+ *
+ * @param   user_char       User input of each character
+ */
+void line_accumlator(uint8_t user_char);
+
+/**
+ * @name process_command
+ *
+ * @brief Performs a Lexical analysis and tokenisis the user's commands
+ * 		  Lookup first token in a table of functions, dispatch to handler function
+ *
+ * @param   input   Character pointer containing the line accumulator input when enter key is pressed 
+ *
+ * @return  void
+ */
+void process_command(char *input);
```

Comment:
I would prefer that this is also not a public function.  Instead of requiring the user to set up their UART/interrupts/callbacks to manually drive these functions, I would prefer that the user is given an `MXC_CLI_Init` function and you handle it for them behind the scenes.